src/parser/ts/index.ts

```diff
@@ -0,0 +1,528 @@
+    serverFilters.map(async filter => {
+      const tsResult = await getTimeseries({ filter }, target, connector, false);
+      if (!tsResult.length) {
+        throw new Error(
+          `No timeseries found for filter ${JSON.stringify(filter)} in expression ${expr}`
+        );
+      }
+      return tsResult;
+    })
+  );
+  const clientFilters = getClientFilters(parsed);
+  const clientFilteredTimeseries = timeseries.map((arr, i) =>
+    applyFilters(arr, clientFilters[i])
+  );
+  const multiaryFuncTsIndices = getIndicesOfMultiaryFunctionArgs(parsed);
+  const permutations = generateAllPossiblePermutations(
+    clientFilteredTimeseries,
+    multiaryFuncTsIndices
+  );
+  const queryExpressions = permutations.map(series => injectTSIdsInExpression(parsed, series));
+  return queryExpressions.map(expression => ({ expression }));
+};
+
+export const injectTSIdsInExpression = (
+  parsedData: STSQueryItem[] | STSFunction,
+  timeseries: TimeSeriesResponseItem[][]
+) => {
+  let i = 0;
+  const exprWithSum = reverse(parsedData, item => {
+    if (isSTSReference(item) && !hasIdsFilter(item)) {
+      const syntheticFilters = item.query.filter(isSTSAggregateFilter);
+      const paramTSIds = timeseries[i++].map(({ id }) => id);
+      const pseudoParsedSTS = paramTSIds.map(id =>
+        STSReference([STSFilter('id', id), ...syntheticFilters])
+      );
+      return pseudoParsedSTS.map(ts => reverse(ts)).join(', ');
+    }
+  });
+  return flattenSumFunctions(exprWithSum);
+};
+
+export const parse = (s: string): STSQueryItem[] | STSFunction => {
+  return parseWith(new Parser(compiledGrammar), s)
+};
+
+export const generateAllPossiblePermutations = <T>(
+  arrayOfArrays: T[][],
+  lockIndices: number[] = []
+): T[][][] => {
+  let res: T[][][] = [];
+  for (let i = 0; i < arrayOfArrays.length; i++) {
+    const locked = lockIndices.includes(i);
+    const array = [...arrayOfArrays[i]];
+    if (res.length) {
+      const temp = [...res];
+      const temptemp = [];
+      for (const t of temp) {
+        if (locked) {
+          temptemp.push([...t, array]);
+        } else {
+          for (const item of array) {
+            temptemp.push([...t, [item]]);
+          }
+        }
+      }
+      res = [...temptemp];
+    } else {
+      if (locked) {
+        res = [[array]];
+      } else {
+        res = array.map(i => [[i]]);
+      }
+    }
+  }
+  return res;
+};
+
+export const convertExpressionToLabel = (
+  expression: string,
+  labelSrc: string,
+  tsMap: TSResponseMap
+) => {
+  const parsed = parse(expression);
+  const labelRes = reverse(parsed, item => {
+    if (isSTSReference(item)) {
+      const [{ value }] = getIdFilters(item);
+      const serie = tsMap[String(value)];
+      if (labelSrc) {
+        return getLabelWithInjectedProps(labelSrc, serie);
+      }
+      return `${serie.name || serie.externalId || serie.id}`;
+    }
+  });
+  return labelRes;
+};
+
+export const getLabelsForExpression = async (
```

Comment:
In this function, we get timeseries again after `getDataQueryRequestItems` and always receive exactly the same data that in `getDataQueryRequestItems`. So all logic regarding labels can be simplified I believe. It's shouldn't be a part of this PR but probably we should keep this in mind

Reply:
yep, agree, in fact I didn't do any smart logic for best performance.
we'll need to review on how many requests we send in this tab, we can lower down number by better logic and in some places - caching

Reply:
oh, caching is smthing that should be also reviewed) But let's have sts perfectly working and deal with this after